test/e2e/lib/template.bash

```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+function fill_in_place_recursively() {
+  local -n key_values=$1 # pass an associate array as a nameref
+  local target_directory=${2:-.}
+  (# use a subshell to expose key-values as variables for envsubst to use
+    for key in "${!key_values[@]}"; do
+      export "$key"="${key_values[$key]}"
+    done
+    while IFS= read -r -d '' file; do
```

Comment:
I think this (L10-17) could use a descriptive comment to not have to decipher the excellent bash wizardry you pull off here if you are just reading the code to get a general understanding.